pyomo/contrib/pynumero/extensions/tests/test_ma27_interface.py

```diff
@@ -0,0 +1,140 @@
+import os
+import ctypes
+from pyomo.contrib.pynumero.dependencies import numpy as np, numpy_available
+if not numpy_available:
+    raise unittest.SkipTest('pynumero MA27 tests require numpy')
+import numpy.ctypeslib as npct
+import pyutilib.th as unittest
+from pyomo.contrib.pynumero.extensions.ma27_interface import *
+
+
+@unittest.skipIf(not MA27Interface.available(), reason='MA27 not available')
+class TestMA27Interface(unittest.TestCase):
+
+    def test_get_cntl(self):
+        ma27 = MA27Interface()
+        self.assertEqual(ma27.get_icntl(1), 6)
+
+        self.assertAlmostEqual(ma27.get_cntl(1), 1e-1) # Numerical pivot threshold
+        self.assertAlmostEqual(ma27.get_cntl(3), 0.0) # Null pivot threshold
+
+    def test_set_icntl(self):
+        ma27 = MA27Interface()
+        ma27.set_icntl(5, 4) # Set output printing to max verbosity
+        ma27.set_icntl(8, 1) # Keep factors when we run out of space
+                             # (so MA27ED can be used)
+        icntl5 = ma27.get_icntl(5)
+        icntl8 = ma27.get_icntl(8)
+        self.assertEqual(icntl5, 4)
+        self.assertEqual(icntl8, 1)
+
+        with self.assertRaisesRegex(TypeError, 'must be an integer'):
+            ma27.set_icntl(1.0, 0)
+        with self.assertRaisesRegex(IndexError, 'is out of range'):
+            ma27.set_icntl(100, 0)
+        with self.assertRaises(ctypes.ArgumentError):
+            ma27.set_icntl(1, 0.0)
+
+    def test_set_cntl(self):
+        ma27 = MA27Interface()
+        ma27.set_cntl(1, 1e-8)
+        ma27.set_cntl(3, 1e-12)
+        self.assertAlmostEqual(ma27.get_cntl(1), 1e-8)
+        self.assertAlmostEqual(ma27.get_cntl(3), 1e-12)
+
+    def test_do_symbolic_factorization(self):
+        ma27 = MA27Interface()
+
+        n = 5
+        ne = 7
+        irn = np.array([1,1,2,2,3,3,5], dtype=np.intc)
+        jcn = np.array([1,2,3,5,3,4,5], dtype=np.intc)
+
+        bad_jcn = np.array([1,2,3,5,3,4], dtype=np.intc)
+
+        ma27.do_symbolic_factorization(n, irn, jcn)
+
+        self.assertEqual(ma27.get_info(1), 0)
+        self.assertEqual(ma27.get_info(5), 14) # Min required num. integer words
+        self.assertEqual(ma27.get_info(6), 20) # Min required num. real words
+
+        with self.assertRaisesRegex(AssertionError, 'Dimension mismatch'):
+            ma27.do_symbolic_factorization(n, irn, bad_jcn)
+
+    def test_do_numeric_factorization(self):
+        ma27 = MA27Interface()
+
+        n = 5
+        ne = 7
+        irn = np.array([1,1,2,2,3,3,5], dtype=np.intc)
+        icn = np.array([1,2,3,5,3,4,5], dtype=np.intc)
+        ent = np.array([2.,3.,4.,6.,1.,5.,1.], dtype=np.double)
+        ent_copy = ent.copy()
+        ma27.do_symbolic_factorization(n, irn, icn)
+
+        status = ma27.do_numeric_factorization(irn, icn, n, ent)
+        self.assertEqual(status, 0)
+
+        expected_ent = [2.,3.,4.,6.,1.,5.,1.,]
+        for i in range(ne):
+            self.assertAlmostEqual(ent_copy[i], expected_ent[i])
```

Comment:
This is a strange test: you are asserting that the values in `ent` haven't changed?  You are comparing the copy of an array you explicitly create with another array you explicitly create...

Reply:
Yes, I am asserting that the double array I passed to MA27 has not changed. This is the array whose memory was getting corrupted before I started memcpy-ing A. The rest of the process, including back solve, proceeded correctly, so this is my only test for such corruption.

Reply:
Right, but shouldn't the test be:
```python
for i in range(ne):
    self.assertAlmostEqual(ent[i], expected_ent[i])
```
(that is, test the numpy array that you passed to ma27?)

Reply:
Oh right, no need to be copying ent here. I added that after my first test failed, and FWIW, this test failed too. Changed